From 89a7c3296d0bb11bb1d5f1f8e204caa77647aff6 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <uwu@icenowy.me>
Date: Sun, 19 Jan 2025 20:49:58 +0800
Subject: [PATCH 6/6] zink: try to merge two VkSubmitInfo's for Imagination
 Rogue

The proprietary Vulkan driver for Imagination Rogue architecture GPUs
has a broken VkQueueSubmit which cannot handle semaphore signaling w/o
command buffers.

Try to merge semaphores signaling submit to command processing one in
this case.

Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
---
 src/gallium/drivers/zink/zink_batch.c  | 9 ++++++++-
 src/gallium/drivers/zink/zink_screen.c | 6 ++++++
 src/gallium/drivers/zink/zink_types.h  | 1 +
 3 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/src/gallium/drivers/zink/zink_batch.c b/src/gallium/drivers/zink/zink_batch.c
index 4d8c18a24ee..5389d66dd46 100644
--- a/src/gallium/drivers/zink/zink_batch.c
+++ b/src/gallium/drivers/zink/zink_batch.c
@@ -746,8 +746,15 @@ submit_queue(void *data, void *gdata, int thread_index)
       );
    }
 
-   if (!si[ZINK_SUBMIT_SIGNAL].signalSemaphoreCount)
+   if (!si[ZINK_SUBMIT_CMDBUF].signalSemaphoreCount &&
+       screen->driver_workarounds.broken_submit) {
+      si[ZINK_SUBMIT_CMDBUF].signalSemaphoreCount = si[ZINK_SUBMIT_SIGNAL].signalSemaphoreCount;
+      si[ZINK_SUBMIT_CMDBUF].pSignalSemaphores = si[ZINK_SUBMIT_SIGNAL].pSignalSemaphores;
+      si[ZINK_SUBMIT_CMDBUF].pNext = si[ZINK_SUBMIT_SIGNAL].pNext;
       num_si--;
+   } else if (!si[ZINK_SUBMIT_SIGNAL].signalSemaphoreCount) {
+      num_si--;
+   }
 
    simple_mtx_lock(&screen->queue_lock);
    VRAM_ALLOC_LOOP(result,
diff --git a/src/gallium/drivers/zink/zink_screen.c b/src/gallium/drivers/zink/zink_screen.c
index 68ae52eb832..10d7a7f2597 100644
--- a/src/gallium/drivers/zink/zink_screen.c
+++ b/src/gallium/drivers/zink/zink_screen.c
@@ -2779,6 +2779,12 @@ init_driver_workarounds(struct zink_screen *screen)
        screen->info.feats.features.geometryShader)
       screen->driver_workarounds.no_linesmooth = true;
 
+   /* Assume Rogue when no geometryShader is available */
+   if (screen->info.driver_props.driverID ==
+       VK_DRIVER_ID_IMAGINATION_PROPRIETARY &&
+       !screen->info.feats.features.geometryShader)
+      screen->driver_workarounds.broken_submit = true;
+
    /* This is a workarround for the lack of
     * gl_PointSize + glPolygonMode(..., GL_LINE), in the imagination
     * proprietary driver.
diff --git a/src/gallium/drivers/zink/zink_types.h b/src/gallium/drivers/zink/zink_types.h
index aa56ab8fdd5..2c7ad622138 100644
--- a/src/gallium/drivers/zink/zink_types.h
+++ b/src/gallium/drivers/zink/zink_types.h
@@ -1566,6 +1566,7 @@ struct zink_screen {
       unsigned z24_unscaled_bias;
       bool broken_const;
       bool broken_demote;
+      bool broken_submit;
    } driver_workarounds;
 };
 
-- 
2.49.0

